File: app/src/test/java/com/tasomaniac/openwith/resolver/HomeScreenResolverPresenterTest.java
package com.tasomaniac.openwith.resolver;

import android.content.Intent;
import android.content.res.Resources;

import com.tasomaniac.openwith.R;

import java.util.Collections;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.*;

public class HomeScreenResolverPresenterTest {

    private static final String TITLE = "title";
    private static final IntentResolver.Data EMPTY_DATA = new IntentResolver.Data(Collections.emptyList(), null, false);

    @Rule public MockitoRule rule = MockitoJUnit.rule();

    @Mock private IntentResolver intentResolver;
    @Mock private ResolverView view;
    @Mock private Resources resources;
    @Mock private Intent sourceIntent;
    @Mock private ResolverNavigation navigation;

    private HomeScreenResolverPresenter presenter;

    @Before
    public void setUp() throws Exception {
        presenter = new HomeScreenResolverPresenter(resources, intentResolver);
        given(intentResolver.getSourceIntent()).willReturn(sourceIntent);

        presenter.bind(view, navigation);
    }

    @Test
    public void unbindShouldNullifyListeners() {
        presenter.unbind(view);

        then(view).should().setListener(null);
        then(intentResolver).should().setListener(null);
    }

    @Test
    public void shouldStartResolvingByDefault() {
        then(intentResolver).should().resolve();
    }

    @Test
    public void givenNonEmptyDataShouldNotifyListenerOnBind() {
        IntentResolver.Data data = mock(IntentResolver.Data.class);
        given(intentResolver.getData()).willReturn(data);

        presenter.bind(view, navigation);

        then(view).should().displayData(data);
    }

    @Test
    public void shouldHaveNoInteractionWithFilteredItem() {
        IntentResolver.Listener listener = captureIntentResolverListener();

        DisplayResolveInfo filteredItem = mock(DisplayResolveInfo.class);
        listener.onIntentResolved(new IntentResolver.Data(Collections.emptyList(), filteredItem, false));

        then(filteredItem).shouldHaveZeroInteractions();
    }

    @Test
    public void givenEmptyResolveListShouldDisplayWarningAndDismiss() {
        IntentResolver.Listener listener = captureIntentResolverListener();
        reset(view);

        listener.onIntentResolved(EMPTY_DATA);

        then(view).should().toast(R.string.empty_resolver_activity);
        then(navigation).should().dismiss();
        then(view).shouldHaveNoMoreInteractions();
    }

    @Test
    public void givenResolveListShouldSetupUI() {
        IntentResolver.Listener listener = captureIntentResolverListener();

        DisplayResolveInfo item = mock(DisplayResolveInfo.class);
        IntentResolver.Data data = dataWithItem(item);
        listener.onIntentResolved(data);

        then(view).should().displayData(data);
    }

    @Test
    public void givenResolveListShouldDisplayTitle() {
        IntentResolver.Listener listener = captureIntentResolverListener();
        given(resources.getString(R.string.add_to_homescreen)).willReturn(TITLE);

        DisplayResolveInfo item = mock(DisplayResolveInfo.class);
        listener.onIntentResolved(dataWithItem(item));

        then(view).should().setTitle(TITLE);
    }

    @Test
    public void shouldReloadWhenPackagedChanged() {
        ResolverView.Listener listener = captureViewListener();
        reset(intentResolver);

        listener.onPackagesChanged();

        then(intentResolver).should().resolve();
    }

    @Test(expected = IllegalStateException.class)
    public void shouldThrowIfActionButtonsClicked() {
        ResolverView.Listener listener = captureViewListener();

        listener.onActionButtonClick(false);
    }

    private IntentResolver.Listener captureIntentResolverListener() {
        ArgumentCaptor<IntentResolver.Listener> argumentCaptor = ArgumentCaptor.forClass(IntentResolver.Listener.class);
        then(intentResolver).should(atLeastOnce()).setListener(argumentCaptor.capture());
        return argumentCaptor.getValue();
    }

    private ResolverView.Listener captureViewListener() {
        ArgumentCaptor<ResolverView.Listener> argumentCaptor = ArgumentCaptor.forClass(ResolverView.Listener.class);
        then(view).should().setListener(argumentCaptor.capture());
        return argumentCaptor.getValue();
    }

    private static IntentResolver.Data dataWithItem(DisplayResolveInfo item) {
        return new IntentResolver.Data(Collections.singletonList(item), null, false);
    }
}

File: app/src/test/java/com/tasomaniac/openwith/resolver/HomeScreenResolverPresenterTest.kt
package com.tasomaniac.openwith.resolver

import android.content.Intent
import android.content.res.Resources
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.atLeastOnce
import com.nhaarman.mockitokotlin2.given
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.reset
import com.nhaarman.mockitokotlin2.then
import com.tasomaniac.openwith.R
import org.junit.Test

class HomeScreenResolverPresenterTest {

    private val intentResolver = mock<IntentResolver>()
    private val view = mock<ResolverView>()
    private val resources = mock<Resources>()
    private val sourceIntent = mock<Intent>()
    private val navigation = mock<ResolverNavigation>()

    private var presenter = HomeScreenResolverPresenter(resources, intentResolver).also {
        it.bind(view, navigation)
    }

    @Test
    fun unbindShouldNullifyListeners() {
        presenter.unbind(view)

        then(view).should().setListener(null)
        then(intentResolver).should().unbind()
    }

    @Test
    fun shouldHaveNoInteractionWithFilteredItem() {
        val listener = captureIntentResolverListener()

        val filteredItem = mock<DisplayActivityInfo>()
        listener.onIntentResolved(
            IntentResolverResult(
                emptyList(),
                filteredItem,
                false
            )
        )

        then(filteredItem).shouldHaveNoInteractions()
    }

    @Test
    fun givenEmptyResolveListShouldDisplayWarningAndDismiss() {
        given(intentResolver.sourceIntent).willReturn(sourceIntent)
        val listener = captureIntentResolverListener()
        reset(view)

        listener.onIntentResolved(EMPTY_RESULT)

        then(view).should().toast(R.string.empty_resolver_activity)
        then(navigation).should().dismiss()
        then(view).shouldHaveNoMoreInteractions()
    }

    @Test
    fun givenResolveListShouldSetupUI() {
        val listener = captureIntentResolverListener()

        val item = mock<DisplayActivityInfo>()
        val result = resultWithItem(item)
        listener.onIntentResolved(result)

        then(view).should().displayData(result)
    }

    @Test
    fun givenResolveListShouldDisplayTitle() {
        val listener = captureIntentResolverListener()
        given(resources.getString(R.string.add_to_homescreen)).willReturn(TITLE)

        val item = mock<DisplayActivityInfo>()
        listener.onIntentResolved(resultWithItem(item))

        then(view).should().setTitle(TITLE)
    }

    @Test
    fun shouldReloadWhenPackagedChanged() {
        val listener = captureViewListener()
        reset(intentResolver)

        listener.onPackagesChanged()

        then(intentResolver).should().resolve()
    }

    @Test(expected = IllegalStateException::class)
    fun shouldThrowIfActionButtonsClicked() {
        val listener = captureViewListener()

        listener.onActionButtonClick(false)
    }

    private fun captureIntentResolverListener() = with(argumentCaptor<IntentResolver.Listener>()) {
        then(intentResolver).should(atLeastOnce()).bind(capture())
        firstValue
    }

    private fun captureViewListener() = with(argumentCaptor<ResolverView.Listener>()) {
        then(view).should().setListener(capture())
        firstValue
    }

    companion object {

        private const val TITLE = "title"
        private val EMPTY_RESULT = IntentResolverResult(emptyList(), null, false)

        private fun resultWithItem(item: DisplayActivityInfo): IntentResolverResult {
            return IntentResolverResult(listOf(item), null, false)
        }
    }
}
